File: aoc/2023/aoc202301.hs
import Data.Char(isDigit,ord)

parse :: String -> Int
parse text
  | null t = 0
  | otherwise = 10*ord (head t) + ord (last t) - 11*ord '0'
  where t = filter isDigit text    

testData :: String
testData = unlines ["1abc2","pqr3stu8vwx","a1b2c3d4e5f","treb7uchet"]

result :: String -> Int
result = sum . map parse . lines

test :: ()
test
  | result testData /= 142 = error "a"
  | otherwise = ()

part1 :: IO Int
part1 = fmap result $ readFile "input/01.txt"

testData2 :: String
testData2 = unlines [
    "two1nine","eightwothree","abcone2threexyz","xtwone3four",
    "4nineeightseven2","zoneight234","7pqrstsixteen"
    ]

-- tricky: numbers can overlap
parse2 :: String -> [Int]
parse2 ('0':rest) = 0 : parse2 rest
parse2 ('z':rest@('e':'r':'o':_)) = 0 : parse2 rest
parse2 ('1':rest) = 1 : parse2 rest
parse2 ('o':rest@('n':'e':_)) = 1 : parse2 rest
parse2 ('2':rest) = 2 : parse2 rest
parse2 ('t':rest@('w':'o':_)) = 2 : parse2 rest
parse2 ('3':rest) = 3 : parse2 rest
parse2 ('t':rest@('h':'r':'e':'e':_)) = 3 : parse2 rest
parse2 ('4':rest) = 4 : parse2 rest
parse2 ('f':rest@('o':'u':'r':_)) = 4 : parse2 rest
parse2 ('5':rest) = 5 : parse2 rest
parse2 ('f':rest@('i':'v':'e':_)) = 5 : parse2 rest
parse2 ('6':rest) = 6 : parse2 rest
parse2 ('s':rest@('i':'x':_)) = 6 : parse2 rest
parse2 ('7':rest) = 7 : parse2 rest
parse2 ('s':rest@('e':'v':'e':'n':_)) = 7 : parse2 rest
parse2 ('8':rest) = 8 : parse2 rest
parse2 ('e':rest@('i':'g':'h':'t':_)) = 8 : parse2 rest
parse2 ('9':rest) = 9 : parse2 rest
parse2 ('n':rest@('i':'n':'e':_)) = 9 : parse2 rest
parse2 (_:rest) = parse2 rest
parse2 [] = []

build2 :: [Int] -> Int
build2 digits | null digits = 0 | otherwise = 10*head digits + last digits

result2 :: String -> Int
result2 = sum . map (build2 . parse2) . lines

test2 :: ()
test2
  | result2 testData2 /= 281 = error "a"
  | otherwise = ()

part2 :: IO Int
part2 = fmap result2 $ readFile "input/01.txt"

File: aoc/2025/aoc202301.go
package main

import (
	"strings"
)

func init() {
	Register(&aoc202301{
		AOC: AOC{
			Day:           1,
			InputFilename: "../2023/input/01.txt",
			Tests: []Test{
				Test{
					Input: `1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet
`,
					Part1: "142",
					Part2: "",
				},
				Test{
					Input: `two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen
`,
					Part1: "",
					Part2: "281",
				},
			},
		},
	})
}

type aoc202301 struct {
	AOC
}

func (aoc *aoc202301) digit(line string) (int, bool) {
	switch line[0] {
	case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		return int(line[0] - '0'), true
	default:
		return 0, false
	}
}

func (aoc *aoc202301) Part1(input *Input) string {
	result := 0
	for line, ok := input.Line(); ok; line, ok = input.Line() {
		tens := 0
		ones := 0
		for line != "" {
			digit, ok := aoc.digit(line)
			line = line[1:]
			if ok {
				tens = digit
				ones = digit
				break
			}
		}
		for line != "" {
			digit, ok := aoc.digit(line)
			line = line[1:]
			if ok {
				ones = digit
			}
		}
		result += tens*10 + ones
	}
	return IntResult(result)
}

func (aoc *aoc202301) digit2(line string) (int, bool) {
	switch line[0] {
	case '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
		return int(line[0] - '0'), true
	case 'e':
		if strings.HasPrefix(line, "eight") {
			return 8, true
		}
	case 'f':
		if strings.HasPrefix(line, "four") {
			return 4, true
		} else if strings.HasPrefix(line, "five") {
			return 5, true
		}
	case 'n':
		if strings.HasPrefix(line, "nine") {
			return 9, true
		}
	case 'o':
		if strings.HasPrefix(line, "one") {
			return 1, true
		}
	case 's':
		if strings.HasPrefix(line, "six") {
			return 6, true
		} else if strings.HasPrefix(line, "seven") {
			return 7, true
		}
	case 't':
		if strings.HasPrefix(line, "two") {
			return 2, true
		} else if strings.HasPrefix(line, "three") {
			return 3, true
		}
	}
	return 0, false
}

func (aoc *aoc202301) Part2(input *Input) string {
	result := 0
	for line, ok := input.Line(); ok; line, ok = input.Line() {
		tens := 0
		ones := 0
		for line != "" {
			digit, ok := aoc.digit2(line)
			line = line[1:]
			if ok {
				tens = digit
				ones = digit
				break
			}
		}
		for line != "" {
			digit, ok := aoc.digit2(line)
			line = line[1:]
			if ok {
				ones = digit
			}
		}
		result += tens*10 + ones
	}
	return IntResult(result)
}
